File: src/risk/sizing.ts
import Decimal from 'decimal.js';

Decimal.set({ precision: 28, rounding: Decimal.ROUND_DOWN });

export interface SizingInput {
  equity: number;
  maxRiskPct: number;
  entryPrice: number;
  stopPrice: number;
  atr14: number;
  minQty: number;
  qtyStep: number;
  maxQty: number;
  maxExposurePct?: number;
  /**
   * Hard ceiling on a single min-lot trade's stop-loss risk, as a fraction of
   * equity. When the risk-budget sizing rounds below the exchange minimum, we
   * still take the trade if min-lot risk stays under this cap. Scales with any
   * balance, so a $10 and a $10k account both "know what to do". Default 0.10.
   */
  maxAbsoluteRiskPct?: number;
}

export interface SizingResult {
  qty: number;
  riskAmount: number;
  notional: number;
  riskPct: number;
}

export function computePositionSize(input: SizingInput): SizingResult {
  const equity = new Decimal(input.equity);
  const maxRiskPct = new Decimal(input.maxRiskPct);
  const entry = new Decimal(input.entryPrice);
  const stop = new Decimal(input.stopPrice);
  const atr = new Decimal(input.atr14);
  const qtyStep = new Decimal(input.qtyStep);
  const minQty = new Decimal(input.minQty);
  const maxQty = new Decimal(input.maxQty);

  const riskBudget = equity.mul(maxRiskPct);
  const riskPerUnit = entry.minus(stop).abs();

  if (riskPerUnit.lte(0)) {
    return { qty: 0, riskAmount: 0, notional: 0, riskPct: 0 };
  }

  // Volatility-scale: reduce size when ATR is large relative to entry
  const atrPct = atr.div(entry);
  const volScale = new Decimal(1).div(new Decimal(1).plus(atrPct.mul(10)));
  const adjustedBudget = riskBudget.mul(volScale);

  let qty = adjustedBudget.div(riskPerUnit);

  // Snap to step
  qty = qty.div(qtyStep).floor().mul(qtyStep);

  // Clamp to maxQty
  if (qty.gt(maxQty)) qty = maxQty;

  // Cap notional at maxExposurePct if provided
  if (input.maxExposurePct) {
    const maxNotional = equity.mul(input.maxExposurePct);
    const notional = qty.mul(entry);
    if (notional.gt(maxNotional)) {
      qty = maxNotional.div(entry).div(qtyStep).floor().mul(qtyStep);
    }
  }

  // Snap to step after caps
  qty = qty.div(qtyStep).floor().mul(qtyStep);

  // Min-lot rounding: when risk-based qty falls below the exchange minimum,
  // round up to minQty only if the resulting stop-loss risk stays within an
  // absolute fraction of equity (default 10%). This scales with any balance —
  // a $10 account trading a cheap coin and a $10k account trading ETH both
  // clear the same proportional bar — while still blocking instruments whose
  // min-lot would risk a dangerous chunk of the account in one trade.
  const maxAbsoluteRiskPct = new Decimal(input.maxAbsoluteRiskPct ?? 0.10);
  if (qty.lt(minQty)) {
    const minLotRisk = minQty.mul(riskPerUnit);
    if (minLotRisk.lte(equity.mul(maxAbsoluteRiskPct))) {
      qty = minQty;
    } else {
      return { qty: 0, riskAmount: 0, notional: 0, riskPct: 0 };
    }
  }

  const finalNotional = qty.mul(entry);
  const finalRisk = qty.mul(riskPerUnit);
  const finalRiskPct = finalRisk.div(equity);

  return {
    qty: qty.toNumber(),
    riskAmount: finalRisk.toNumber(),
    notional: finalNotional.toNumber(),
    riskPct: finalRiskPct.toNumber(),
  };
}

File: bybit_agent/risk/sizing.py
"""Position sizing — ports src/risk/sizing.ts using stdlib Decimal (prec 28, ROUND_DOWN)
to match decimal.js exactly. Risk-budget sizing, volatility scaling, step snapping,
exposure cap, and min-lot rounding bounded by an absolute fraction of equity.

Leverage changes the *notional exposure* available but not the *risk budget*:
  risk_budget  = equity × maxRiskPct        (how much of real equity you can lose)
  max_notional = equity × maxExposurePct    (real buying-power cap; leverage applied below)
  qty          = risk_budget / risk_per_unit

With leverage L, the exchange lets you hold L× the position your margin supports.
The exposure cap accounts for this: max_notional_leveraged = max_notional × leverage.
The absolute-risk rescue also uses ACTUAL equity (not leveraged) so min-lot
orders never risk more than maxAbsoluteRiskPct × real equity.
"""

from __future__ import annotations

from dataclasses import dataclass
from decimal import ROUND_DOWN, Decimal, getcontext

getcontext().prec = 28
getcontext().rounding = ROUND_DOWN

_D = Decimal


@dataclass
class SizingInput:
    equity: float
    maxRiskPct: float
    entryPrice: float
    stopPrice: float
    atr14: float
    minQty: float
    qtyStep: float
    maxQty: float
    leverage: int = 1                   # exchange leverage (1 = no leverage)
    maxExposurePct: float | None = None
    maxAbsoluteRiskPct: float | None = None


@dataclass
class SizingResult:
    qty: float
    riskAmount: float
    notional: float
    riskPct: float


def _floor_step(qty: Decimal, step: Decimal) -> Decimal:
    return (qty / step).to_integral_value(rounding=ROUND_DOWN) * step


def compute_position_size(inp: SizingInput) -> SizingResult:
    equity    = _D(str(inp.equity))
    max_risk  = _D(str(inp.maxRiskPct))
    entry     = _D(str(inp.entryPrice))
    stop      = _D(str(inp.stopPrice))
    atr       = _D(str(inp.atr14))
    qty_step  = _D(str(inp.qtyStep))
    min_qty   = _D(str(inp.minQty))
    max_qty   = _D(str(inp.maxQty))
    leverage  = _D(str(max(1, inp.leverage)))

    risk_budget  = equity * max_risk
    risk_per_unit = abs(entry - stop)

    if risk_per_unit <= 0:
        return SizingResult(0, 0, 0, 0)

    atr_pct    = atr / entry if entry > 0 else _D(0)
    vol_scale  = _D(1) / (_D(1) + atr_pct * _D(10))
    adj_budget = risk_budget * vol_scale

    qty = adj_budget / risk_per_unit
    qty = _floor_step(qty, qty_step)

    if qty > max_qty:
        qty = max_qty

    # Exposure cap uses LEVERAGED notional: with 10× the exchange lets you hold
    # 10× what your margin supports.  Cap is still anchored to real equity so
    # we never over-size, even with high leverage.
    if inp.maxExposurePct:
        max_notional = equity * _D(str(inp.maxExposurePct)) * leverage
        notional = qty * entry
        if notional > max_notional:
            qty = _floor_step(max_notional / entry, qty_step)

    qty = _floor_step(qty, qty_step)

    # Min-lot rescue: absolute risk check uses REAL equity (not leveraged).
    max_abs = _D(str(inp.maxAbsoluteRiskPct if inp.maxAbsoluteRiskPct is not None else 0.10))
    if qty < min_qty:
        min_lot_risk = min_qty * risk_per_unit
        if min_lot_risk <= equity * max_abs:
            qty = min_qty
        else:
            return SizingResult(0, 0, 0, 0)

    final_notional = qty * entry
    final_risk     = qty * risk_per_unit
    final_risk_pct = final_risk / equity   # expressed as fraction of REAL equity

    return SizingResult(
        qty=float(qty),
        riskAmount=float(final_risk),
        notional=float(final_notional),
        riskPct=float(final_risk_pct),
    )

